File: back/app/sockets/presence.py
"""Учёт присутствия (онлайн-статус) пользователей.

Держим в памяти процесса состояние каждого сокет-соединения: к какому
пользователю относится, видна ли вкладка и когда мы последний раз слышали
от клиента heartbeat. Пользователь «онлайн», пока у него есть хотя бы одно
соединение с видимой вкладкой И недавним heartbeat'ом.

Почему heartbeat, а не одна только видимость: на мобильных (особенно iOS
Safari) при сворачивании приложения/блокировке экрана сокет «замораживается»
— дисконнект приходит с большой задержкой или теряется, и сервер долго
считает пользователя онлайн. Клиент шлёт `presence:heartbeat` каждые ~25с,
пока вкладка видима. Фоновая задача `sweep_stale` ходит раз в `SWEEP_INTERVAL`
секунд и помечает «не в сети» тех, от кого heartbeat'а не было дольше
`STALE_AFTER` секунд. Так last_seen всегда близок к реальному уходу.

Развёртывание — один app-контейнер с eventlet, поэтому in-memory достаточно.
Если когда-нибудь появится несколько процессов socketio, presence надо будет
вынести в Redis.
"""
from datetime import datetime, timezone
from time import monotonic

from app.extensions import db, socketio
from app.models import User
from app.utils.logger import get_logger

logger = get_logger(__name__)

# Интервал между прогонами sweep_stale (сек).
SWEEP_INTERVAL = 15.0
# Если от sid нет heartbeat'а дольше — считаем «вкладка ушла в фон».
STALE_AFTER = 60.0

# sid -> user_id
_sid_user: dict[str, int] = {}
# sid -> видна ли вкладка этого соединения (по умолчанию True на connect)
_sid_visible: dict[str, bool] = {}
# sid -> monotonic time последнего heartbeat / события активности
_sid_last_beat: dict[str, float] = {}
# user_id'ы, считающиеся сейчас онлайн (есть хотя бы одна видимая вкладка)
_online: set[int] = set()

_sweep_started = False


def _has_visible_connection(user_id: int) -> bool:
    return any(
        _sid_visible.get(sid, False)
        for sid, uid in _sid_user.items()
        if uid == user_id
    )


def has_any_connection(user_id: int) -> bool:
    """Есть ли у пользователя хоть одно живое сокет-соединение (видимое или
    нет). Для звонков важно именно это: при перезагрузке вкладки/смене сети
    соединение на мгновение рвётся и переустанавливается, но в фоне сокет
    тоже валиден — звонок не должен завершаться, пока есть любой коннект."""
    return any(uid == user_id for uid in _sid_user.values())


def _persist_last_seen(user_id: int) -> datetime:
    now = datetime.now(timezone.utc)
    try:
        db.session.execute(
            db.update(User).where(User.id == user_id).values(last_seen_at=now)
        )
        db.session.commit()
    except Exception as e:  # noqa: BLE001
        db.session.rollback()
        logger.warning("presence.last_seen_failed",
                       extra={"extra": {"user_id": user_id, "error": str(e)}})
    return now


def _set_online(user_id: int, online: bool) -> None:
    """Меняем онлайн-статус пользователя и рассылаем событие только на переходе,
    чтобы не спамить presence:update и не «двигать» last_seen вперёд при поздних
    дисконнектах уже ушедшего в фон пользователя."""
    if online:
        if user_id in _online:
            return
        _online.add(user_id)
        socketio.emit("presence:update",
                      {"user_id": user_id, "online": True, "last_seen_at": None},
                      room="all")
    else:
        if user_id not in _online:
            return
        _online.discard(user_id)
        now = _persist_last_seen(user_id)
        socketio.emit("presence:update",
                      {"user_id": user_id, "online": False, "last_seen_at": now.isoformat()},
                      room="all")


def on_connect(sid: str, user_id) -> None:
    user_id = int(user_id)
    _sid_user[sid] = user_id
    _sid_visible[sid] = True
    _sid_last_beat[sid] = monotonic()
    _set_online(user_id, True)


def on_disconnect(sid: str) -> None:
    user_id = _sid_user.pop(sid, None)
    _sid_visible.pop(sid, None)
    _sid_last_beat.pop(sid, None)
    if user_id is None:
        return
    if not _has_visible_connection(user_id):
        _set_online(user_id, False)


def on_visibility(sid: str, visible: bool) -> None:
    """Клиент сообщил, что его вкладка стала видимой/скрытой."""
    if sid not in _sid_user:
        return
    _sid_visible[sid] = bool(visible)
    if visible:
        _sid_last_beat[sid] = monotonic()
    user_id = _sid_user[sid]
    _set_online(user_id, _has_visible_connection(user_id))


def on_heartbeat(sid: str) -> None:
    """Регулярный пинг от клиента. Подтверждает, что вкладка жива и видима."""
    if sid not in _sid_user:
        return
    _sid_last_beat[sid] = monotonic()
    # Если sid когда-то был помечен как невидимый из-за просрочки heartbeat'а,
    # heartbeat возвращает его в строй.
    if not _sid_visible.get(sid, False):
        _sid_visible[sid] = True
        user_id = _sid_user[sid]
        _set_online(user_id, True)


def _sweep_once(app) -> None:
    """Один прогон: помечаем все sid'ы, от которых давно не было heartbeat'а,
    как невидимые. Когда у пользователя не остаётся видимых соединений —
    выставляем offline с актуальным last_seen."""
    now = monotonic()
    affected_users: set[int] = set()
    for sid, last in list(_sid_last_beat.items()):
        if not _sid_visible.get(sid, False):
            continue
        if now - last > STALE_AFTER:
            _sid_visible[sid] = False
            uid = _sid_user.get(sid)
            if uid is not None:
                affected_users.add(uid)
    if not affected_users:
        return
    with app.app_context():
        for uid in affected_users:
            if not _has_visible_connection(uid):
                _set_online(uid, False)


def start_sweeper(app) -> None:
    """Запускаем фоновую задачу один раз на процесс. Использует socketio
    background-task — eventlet-совместимый sleep."""
    global _sweep_started
    if _sweep_started:
        return
    _sweep_started = True

    def _loop():
        while True:
            try:
                _sweep_once(app)
            except Exception as e:  # noqa: BLE001
                logger.warning("presence.sweep_failed", extra={"extra": {"error": str(e)}})
            socketio.sleep(SWEEP_INTERVAL)

    socketio.start_background_task(_loop)


def online_user_ids() -> list[int]:
    return list(_online)

File: back-go/gateway/internal/presence/presence.go
// Package presence — учёт присутствия (онлайн-статус) пользователей.
//
// Порт прежнего back/app/sockets/presence.py с переносом состояния в Redis —
// это снимает ограничение «один процесс»: всё состояние соединений живёт
// в общих ключах, событие presence:update уходит через Redis-канал шлюза
// и доставляется клиентам любого инстанса.
//
// Пользователь «онлайн», пока у него есть хотя бы одно соединение с видимой
// вкладкой И недавним heartbeat'ом. Почему heartbeat, а не одна видимость:
// на мобильных (особенно iOS Safari) при сворачивании сокет «замораживается»
// — дисконнект приходит с большой задержкой или теряется. Клиент шлёт
// presence:heartbeat каждые ~25с, пока вкладка видима; sweeper раз в
// SweepInterval опускает в офлайн тех, от кого сигналов не было дольше
// StaleAfter. last_seen_at пишется в users на переходе в офлайн.
//
// Ключи Redis:
//   gw2:presence:beats  — ZSET, member "uid:connID", score — unix-время
//                         последнего сигнала живой видимой вкладки;
//   gw2:presence:online — SET онлайн-пользователей (для переходов и REST).
package presence

import (
	"context"
	"fmt"
	"log/slog"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/redis/go-redis/v9"
)

const (
	beatsKey  = "gw2:presence:beats"
	onlineKey = "gw2:presence:online"

	// SweepInterval / StaleAfter — те же, что во Flask-presence.
	SweepInterval = 15 * time.Second
	StaleAfter    = 60 * time.Second

	roomAll = "all"
)

// LastSeenWriter — запись users.last_seen_at (pgx-пул).
type LastSeenWriter interface {
	SetLastSeen(ctx context.Context, userID int64, at time.Time) error
}

// PGLastSeen — стандартная реализация поверх общей PostgreSQL.
type PGLastSeen struct{ Pool *pgxpool.Pool }

func (w PGLastSeen) SetLastSeen(ctx context.Context, userID int64, at time.Time) error {
	_, err := w.Pool.Exec(ctx,
		`UPDATE users SET last_seen_at = $2 WHERE id = $1`, userID, at)
	return err
}

// Bus — публикация presence:update (events.Publisher gateway-канала).
type Bus interface {
	Publish(ctx context.Context, event string, rooms []string, payload any)
}

type Presence struct {
	rdb      *redis.Client
	lastSeen LastSeenWriter
	bus      Bus
	log      *slog.Logger
	now      func() time.Time
}

func New(rdb *redis.Client, lastSeen LastSeenWriter, bus Bus, log *slog.Logger) *Presence {
	return &Presence{rdb: rdb, lastSeen: lastSeen, bus: bus, log: log, now: time.Now}
}

func member(userID int64, connID string) string {
	return strconv.FormatInt(userID, 10) + ":" + connID
}

func memberUserID(m string) int64 {
	idx := strings.IndexByte(m, ':')
	if idx <= 0 {
		return 0
	}
	id, _ := strconv.ParseInt(m[:idx], 10, 64)
	return id
}

// isoUTC — datetime.isoformat() Python: микросекунды опускаются, если 0.
func isoUTC(t time.Time) string {
	u := t.UTC()
	s := u.Format("2006-01-02T15:04:05")
	if us := u.Nanosecond() / 1000; us != 0 {
		s += fmt.Sprintf(".%06d", us)
	}
	return s + "+00:00"
}

// beat — пометить соединение живым и видимым.
func (p *Presence) beat(ctx context.Context, userID int64, connID string) {
	if err := p.rdb.ZAdd(ctx, beatsKey, redis.Z{
		Score:  float64(p.now().Unix()),
		Member: member(userID, connID),
	}).Err(); err != nil {
		p.log.Warn("presence.beat_failed", "user_id", userID, "error", err)
		return
	}
	p.setOnline(ctx, userID)
}

// drop — соединение больше не видимо/живо.
func (p *Presence) drop(ctx context.Context, userID int64, connID string) {
	if err := p.rdb.ZRem(ctx, beatsKey, member(userID, connID)).Err(); err != nil {
		p.log.Warn("presence.drop_failed", "user_id", userID, "error", err)
		return
	}
	p.maybeOffline(ctx, userID)
}

func (p *Presence) OnConnect(ctx context.Context, userID int64, connID string) {
	p.beat(ctx, userID, connID)
}

func (p *Presence) OnDisconnect(ctx context.Context, userID int64, connID string) {
	p.drop(ctx, userID, connID)
}

// OnVisibility — клиент сообщил, что его вкладка стала видимой/скрытой.
func (p *Presence) OnVisibility(ctx context.Context, userID int64, connID string, visible bool) {
	if visible {
		p.beat(ctx, userID, connID)
	} else {
		p.drop(ctx, userID, connID)
	}
}

// OnHeartbeat — регулярный пинг живой видимой вкладки (возвращает в строй
// соединение, опущенное sweeper'ом).
func (p *Presence) OnHeartbeat(ctx context.Context, userID int64, connID string) {
	p.beat(ctx, userID, connID)
}

// setOnline — переход в онлайн; событие только на переходе, чтобы не спамить
// presence:update.
func (p *Presence) setOnline(ctx context.Context, userID int64) {
	added, err := p.rdb.SAdd(ctx, onlineKey, userID).Result()
	if err != nil {
		p.log.Warn("presence.online_failed", "user_id", userID, "error", err)
		return
	}
	if added == 0 {
		return
	}
	p.bus.Publish(ctx, "presence:update", []string{roomAll}, map[string]any{
		"user_id": userID, "online": true, "last_seen_at": nil,
	})
}

// maybeOffline — если живых видимых соединений не осталось, выставить офлайн
// с актуальным last_seen.
func (p *Presence) maybeOffline(ctx context.Context, userID int64) {
	alive, err := p.aliveUsers(ctx)
	if err != nil {
		p.log.Warn("presence.alive_failed", "error", err)
		return
	}
	if alive[userID] {
		return
	}
	p.setOffline(ctx, userID)
}

func (p *Presence) setOffline(ctx context.Context, userID int64) {
	removed, err := p.rdb.SRem(ctx, onlineKey, userID).Result()
	if err != nil {
		p.log.Warn("presence.offline_failed", "user_id", userID, "error", err)
		return
	}
	if removed == 0 {
		return
	}
	now := p.now().UTC()
	if err := p.lastSeen.SetLastSeen(ctx, userID, now); err != nil {
		p.log.Warn("presence.last_seen_failed", "user_id", userID, "error", err)
	}
	p.bus.Publish(ctx, "presence:update", []string{roomAll}, map[string]any{
		"user_id": userID, "online": false, "last_seen_at": isoUTC(now),
	})
}

// aliveUsers — пользователи с хотя бы одним свежим видимым соединением.
func (p *Presence) aliveUsers(ctx context.Context) (map[int64]bool, error) {
	minScore := strconv.FormatInt(p.now().Add(-StaleAfter).Unix(), 10)
	members, err := p.rdb.ZRangeByScore(ctx, beatsKey, &redis.ZRangeBy{
		Min: "(" + minScore, Max: "+inf",
	}).Result()
	if err != nil {
		return nil, err
	}
	out := map[int64]bool{}
	for _, m := range members {
		if id := memberUserID(m); id > 0 {
			out[id] = true
		}
	}
	return out, nil
}

// OnlineUserIDs — снимок онлайн-пользователей (REST /api/messenger/presence).
func (p *Presence) OnlineUserIDs(ctx context.Context) ([]int64, error) {
	raw, err := p.rdb.SMembers(ctx, onlineKey).Result()
	if err != nil {
		return nil, err
	}
	out := make([]int64, 0, len(raw))
	for _, s := range raw {
		if id, err := strconv.ParseInt(s, 10, 64); err == nil {
			out = append(out, id)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i] < out[j] })
	return out, nil
}

// SweepOnce — один прогон: вычистить просроченные соединения и опустить в
// офлайн пользователей, у которых не осталось живых видимых вкладок
// (включая «осиротевших» после рестарта шлюза).
func (p *Presence) SweepOnce(ctx context.Context) {
	maxScore := strconv.FormatInt(p.now().Add(-StaleAfter).Unix(), 10)
	if err := p.rdb.ZRemRangeByScore(ctx, beatsKey, "-inf", maxScore).Err(); err != nil {
		p.log.Warn("presence.sweep_trim_failed", "error", err)
		return
	}
	alive, err := p.aliveUsers(ctx)
	if err != nil {
		p.log.Warn("presence.sweep_failed", "error", err)
		return
	}
	online, err := p.OnlineUserIDs(ctx)
	if err != nil {
		p.log.Warn("presence.sweep_failed", "error", err)
		return
	}
	for _, uid := range online {
		if !alive[uid] {
			p.setOffline(ctx, uid)
		}
	}
}

// RunSweeper — фоновый цикл; завершается по ctx.
func (p *Presence) RunSweeper(ctx context.Context) {
	ticker := time.NewTicker(SweepInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			p.SweepOnce(ctx)
		}
	}
}
